### ISSUE TITLE
Fallback Dispatcher Not Working As Expected

### ISSUE THREAD

**Issue by timchase01** [microcks/microcks#741]
Hey All,

I'm trying to get the Fallback Dispatcher working and for the life of me I can't get it to work.  I installed Microcks using Podman on a Linux server.  I believe I installed the 1.6.2 snapshot, or at least that is what the pom.xml indicates it is using.  I used the Car Open API example and loaded that into Apicurio and then used Apicurio to push it over to Microcks.  Then altered the Dispatcher to Fallback from the default that Microcks originally set.

The image below shows what I have as the Dispatcher.  The problem is that no matter what I do, I always get the Not_Found response, even if sending a valid mocked Manufacturer name.   If I switch back to URI_PARTS instead of FALLBACK then it works as expected.   What am I doing wrong?

![image](https://user-images.githubusercontent.com/53839570/214164250-d0581489-f0ff-4dcb-942c-7f2614ab2918.png)
![image](https://user-images.githubusercontent.com/53839570/214164767-2f0458c0-1d4c-4ac4-8070-c2f1e382c122.png)

**Comment by lbroudoux** [comment]
Hey @timchase01,

I had a look at your case and though I didn't strictly reproduced it, I found a bug that may have the same behaviour as what you're seeing. Let me explain.

Changing to `FALLBACK` with you config works well if I change it after a first import and doesn't re-import the OpenAPI file. On subsequent imports, Microcks is not recomputing the different paths corresponding to different examples and always respond a 404 code (that you may interpret as your `Not_Found` message).

Can you check this and confirm:
* that `FALLBACK` is actually working after a clean, fresh, first import,
* the `FALLBACK` stops working and mock answers 404 after a new import on top of existing api/service in Microcks.

Thanks,

**Comment by lbroudoux** [comment]
I've referenced it as a new issue in #742

**Comment by lbroudoux** [comment]
Thanks for the OpenAPI file!

In your fallback definition, if you replace `URI_ELEMENTS` by `URI_PARTS` it should work. Made the change on my side and looks like it is working.

```sh
$ curl http://localhost:8080/rest/OpenAPI+Car+API/1.1/manufacturer/Opel/cars -v
*   Trying 127.0.0.1:8080...
* Connected to localhost (127.0.0.1) port 8080 (#0)
> GET /rest/OpenAPI+Car+API/1.1/manufacturer/Opel/cars HTTP/1.1
> Host: localhost:8080
> User-Agent: curl/7.85.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 404 
< Vary: Origin
< Vary: Access-Control-Request-Method
< Vary: Access-Control-Request-Headers
< Content-Type: text/html;charset=UTF-8
< Content-Length: 9
< Date: Tue, 24 Jan 2023 16:35:57 GMT
< 
* Connection #0 to host localhost left intact
Not Found
```

**Comment by timchase01** [comment]
Cool, I'll try that now.  Also, just as an FYI, I am also seeing the bug you found as well.

**Comment by timchase01** [comment]
Hum, I changed to URI_PARTS and pushed a new version to Microcks, but still getting the Not_Found response instead of the Dodge response.  Also, I had it set as URI_ELEMENTS as that is what Microcks originally pulled it in as before I changed it to FALLBACK in previous instances.  In case it matters.

http://microcks.ftr.com:8080/rest/OpenAPI+Car+API/1.2/manufacturer/Dodge/cars

16:45:16.244  INFO 1 --- [080-exec-4] i.github.microcks.web.RestController     : Servicing mock response for service [OpenAPI+Car+API, 1.2] on uri /rest/OpenAPI+Car+API/1.2/manufacturer/Dodge/cars with verb GET
16:45:16.244 DEBUG 1 --- [080-exec-4] i.github.microcks.web.RestController     : Request body: null
16:45:16.244 DEBUG 1 --- [080-exec-4] i.github.microcks.web.RestController     : Found resourcePath: /manufacturer/Dodge/cars
16:45:16.246 DEBUG 1 --- [080-exec-4] i.github.microcks.web.RestController     : Found a valid operation GET /manufacturer/{name}/cars with rules: {"dispatcher":"URI_PARTS","dispatcherRules":"name","fallback":"Not_Found"}
16:45:16.246 DEBUG 1 --- [080-exec-4] i.github.microcks.web.RestController     : Dispatch criteria for finding response is /name=Dodge
16:45:16.249 DEBUG 1 --- [080-exec-4] i.g.m.web.MockControllerCommons          : Mock delay is turned on, waiting if necessary...
16:45:16.249 DEBUG 1 --- [080-exec-4] i.g.m.web.MockControllerCommons          : Delay now expired, releasing response !
16:45:16.249 DEBUG 1 --- [080-exec-4] i.g.m.web.MockControllerCommons          : Mock invocation event has been published
16:45:16.249 DEBUG 1 --- [  task-520] i.g.m.listener.DailyStatisticsFeeder     : Received a MockInvocationEvent on OpenAPI Car API - v1.2
16:45:16.249 DEBUG 1 --- [  task-520] i.g.m.listener.DailyStatisticsFeeder     : hourKey for statistic is 16
16:45:16.249 DEBUG 1 --- [  task-520] i.g.m.listener.DailyStatisticsFeeder     : minuteKey for statistic is 1005
16:45:16.250 DEBUG 1 --- [  task-520] i.g.m.listener.DailyStatisticsFeeder     : There's no statistics for 20230124 yet. Create one....
16:45:16.252 DEBUG 1 --- [  task-520] i.g.m.listener.DailyStatisticsFeeder     : Processing of MockInvocationEvent done !

**Comment by lbroudoux** [comment]
One thing you can check though is the completion of operation `resourcePaths`. Open the developer tools in your browser and check the JSON objects retrieved by UI. Expand the correct operation and check its `resourcePaths` array as show below:

<img width="1374" alt="Capture d’écran 2023-01-24 à 18 08 10" src="https://user-images.githubusercontent.com/1538635/214360438-349f0c60-7f13-46d8-b5d8-433f53f40955.png">

You can also check the `dispatchCriteria` property of the `Dodge` response in the `messagesMap` from the same panel. You saw in the log that its value must be `/name=Dodge`:

<img width="1278" alt="Capture d’écran 2023-01-24 à 18 10 56" src="https://user-images.githubusercontent.com/1538635/214361249-68fd1abc-21f8-4317-9534-36ded43d7f1b.png">

**Comment by lbroudoux** [comment]
I'm currently running the `1.7.0-SNAPSHOT` version but didn't think is has changed lately (apart the bug I've fixed few hours ago 😉 ). The version I'm using is in `quay.io/microcks/microcks:next` container image. Let me know the version you're using so that I'll check with this one.

**Comment by lbroudoux** [comment]
If you installed from `master` branch it should be `1.6.1`. Check the about box in UI

**Comment by timchase01** [comment]
Yup, 1.6.1

**Comment by lbroudoux** [comment]
OK got it ... I reproduce your behaviour on `1.6.1`... my bad ... 😖

It appears that the bug I identified lately also occurs when `FALLBACK` is also specified using `x-microcks-operation` : the `resourcePaths` are ignored and prevent the root dispatcher from fallback specification to work correctly.

I had a look at my existing test cases to check why I didn't meet it before and it appears that I only specify Fallback using an [`APIMetadata` document](https://microcks.io/documentation/using/advanced/metadata/).

So the good news is that fix is already present but you have to switch to `quay.io/microcks/microcks:next` container image. Is it easily doable for you?

**Comment by timchase01** [comment]
I did validate the items you mentioned in the dev tools.

`2: {name: "GET /manufacturer/{name}/cars", method: "GET", dispatcher: "FALLBACK",…}
    defaultDelay: 0
    dispatcher: "FALLBACK"
    dispatcherRules: "{\"dispatcher\":\"URI_PARTS\",\"dispatcherRules\":\"name\",\"fallback\":\"Not_Found\"}"
    method: "GET"
    name: "GET /manufacturer/{name}/cars"
    resourcePaths: ["/manufacturer/{name}/cars"]
        0: "/manufacturer/{name}/cars"`

The difference being that in your resourcePaths it shows all of the different valid resource paths for each of the valid responses.   For mine I just have the one showed above.

**Comment by lbroudoux** [comment]
Yep. This prevent the root `URI_PARTS` dispatcher to work correctly...

**Comment by timchase01** [comment]
Sorry missed your above comment about your bug fix will have fixed it.  Sure let me work on getting the latest built deployed.  Appreciate the help!

**Comment by timchase01** [comment]
Sorry, another question for you.  I'm trying to use the helm chart to install it and updated the YAML to have quay.io/microcks/microcks:next as the image for Microcks.  I left everything else the same.  The deploy failed though.

**Comment by timchase01** [comment]
Just wanted to let you know that I finally got the :next version installed and things are working great now.  Appreciate it.

**Comment by timchase01** [comment]
@lbroudoux, is there by chance a way that you can change the Operator install so that it makes both the Latest (1.6.1) available as well as the :next branch?   I was on a call with RedHat today trying to get it to install into OpenShift trying both the Helm Chart and the Operator.  With the helm chart we tried change it for :1.6.1 to :next but kept getting a gateway timeout error.  They said the preferred method would be to use operations anyway, but they weren't able to trick it into installing the :next container and suggested I reach out to you to ask as it would be the preferred way to handle it.

**Comment by lbroudoux** [comment]
Hey @timchase01 
The operator for installing `next` (soon to become `1.7.0`) will be released simultaneously to the `1.7.0` (I hope within a week or two). In the meantime, you can try deploying the operator using the `quay.io/microcks/microcks-ansible-operator:nightly` that should be able to realize a correct setup when specifying `next` as the version into `MicrocksInstall` custom resource.
Stay tuned for release!

**Comment by lbroudoux** [comment]
I'm closing this one and please don't hesitate to open a new one at github.com/microcks/microcks-ansible-operator if you're facing issues with the latest version.
